ocaml/xapi/xapi_pool.ml

```diff
@@ -3605,3 +3656,223 @@ let set_https_only ~__context ~self:_ ~value =
         )
         (Db.Host.get_all ~__context)
   )
+
+let install_rpmgpgkey ~__context ~self:_ ~name ~pubkey =
+  (* This function runs on the coordinator host. *)
+  let open Gpg in
+  assert_name_is_valid ~name ;
+  let md = parse_pubkey ~pubkey in
+  let expr =
+    let open Db_filter_types in
+    Eq (Field "type", Literal "rpm_pubkey")
+  in
+  ( Db.Gpg_key.get_refs_where ~__context ~expr
+  |> List.filter_map (fun ref ->
+         let name_in_db = Db.Gpg_key.get_name ~__context ~self:ref in
+         let fp_in_db = Db.Gpg_key.get_fingerprint ~__context ~self:ref in
+         match (name, md.PubKeyMetaData.fingerprint) with
+         | n, fp when n = name_in_db && fp = fp_in_db ->
+             debug "RPM GPG public key %s with fingerprint %s exists already" n
+               fp ;
+             Some ref
+         | n, _ when n = name_in_db ->
+             raise
+               Api_errors.(
+                 Server_error
+                   (rpm_gpg_key_record_with_same_name_already_exists, [name])
+               )
+         | _, fp when fp = fp_in_db ->
+             raise
+               Api_errors.(
+                 Server_error
+                   ( rpm_gpg_key_with_same_fingerprint_already_exists
+                   , [name_in_db]
+                   )
+               )
+         | _ ->
+             None
+     )
+  |> fun gpg_key_refs ->
+    (* Install the key on the coordinator host firstly. *)
+    Rpm_gpg_key.install_rpmgpgkey ~name ~pubkey
+      ~fingerprint:md.PubKeyMetaData.fingerprint ;
+    match gpg_key_refs with
```

Comment:
I believe the key should be installed in the filesystem only if the database record is missing, otherwise we're writing to the filesystem needlessly

Reply:
I think this would correct the case that the key file on coordinator is missing.